vmm/src/device_manager.rs

```diff
@@ -428,38 +430,31 @@ impl DeviceManager {
         let kvm_gsi_msi_routes: Arc<Mutex<HashMap<u32, KvmRoutingEntry>>> =
             Arc::new(Mutex::new(HashMap::new()));
 
-        // Here we create a first interrupt manager that will be directly
-        // passed down to the Ioapic. The reason we need this extra interrupt
-        // manager is because the more global one will need a handler onto the
-        // Ioapic itself. We didn't want to solve this problem by adding some
-        // setter to the KvmInterruptManager as this would have required the
-        // interrupt manager to be mutable.
-        //
-        // One thing to note, it is safe to create two interrupt managers
-        // without risking some concurrency between the two since the list
-        // of GSI routes is shared and protected by a Mutex.
-        let ioapic_interrupt_manager: Arc<dyn InterruptManager> =
-            Arc::new(KvmInterruptManager::new(
+        // First we create the MSI interrupt manager, the legacy one is created
+        // later, after the IOAPIC device creation.
+        // The reason we create the MSI one first is because the IOAPIC needs it,
+        // and then the legacy interrupt manager needs an IOAPIC. So we're
+        // handling a linear dependency chain:
+        // msi_interrupt_manager <- IOAPIC <- legacy_interrupt_manager.
+        let msi_interrupt_manager: Arc<dyn InterruptManager> =
+            Arc::new(KvmMsiInterruptManager::new(
                 Arc::clone(&address_manager.allocator),
-                vm_fd.clone(),
+                vm_fd,
                 Arc::clone(&kvm_gsi_msi_routes),
-                None,
             ));
 
-        let ioapic = DeviceManager::add_ioapic(&address_manager, ioapic_interrupt_manager)?;
+        let ioapic =
+            DeviceManager::add_ioapic(&address_manager, Arc::clone(&msi_interrupt_manager))?;
 
         // Creation of the global interrupt manager, which can take a hold onto
         // the brand new Ioapic.
         //
         // Note the list of GSI routes is Arc cloned, the same way it was Arc
         // cloned for the interrupt manager dedicated to the Ioapic. That's how
         // both interrupt managers are going to share the list correctly.
-        let interrupt_manager: Arc<dyn InterruptManager> = Arc::new(KvmInterruptManager::new(
-            Arc::clone(&address_manager.allocator),
-            vm_fd,
-            Arc::clone(&kvm_gsi_msi_routes),
-            Some(ioapic.clone()),
-        ));
+        let legacy_interrupt_manager: Arc<dyn InterruptManager> = Arc::new(
```

Comment:
Could you update the comment here?

Reply:
I think it's already updated, see:
https://github.com/cloud-hypervisor/cloud-hypervisor/pull/704/files#diff-69cff8b0e315f40c982444fa56351aa6R434-R441

Let me know if you had something else in mind.

Reply:
Well I saw you've updated the other comment, but you could also update this one since it says `Creation of the global interrupt manager` when it really is the creation of the legacy interrupt manager. Just a small nit.

Reply:
A good nit. Fixed, thanks.